### ISSUE TITLE
[Bug]: Optional chaining unnecessarily memoized 

### ISSUE THREAD

**Issue by Austaras** [babel/babel#13956]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
foo?.bar()
```

### Current and expected behavior

Currently compiled to
```js
var _foo;

(_foo = foo) == null ? void 0 : _foo.bar();
```
but should be compiled to
```js
foo == null ? void 0 : foo.bar();
```

### Environment

babel playground

**Comment by JLHwung** [comment]
This is expected. Accessing global variable can have side effect:

```js
var counter = 0;
Object.assign(globalThis, {
  get foo() {
    counter++;
    return { bar() {} }
  }
}
foo?.bar();
expect(counter).toBe(1);
```

**Comment by Austaras** [comment]
That would be a much rare case. So should babel add another assumption like 'pureGlobalThisGetter' or just assume global getter is pure when 'pureGetter' is set?

**Comment by JLHwung** [comment]
Respecting the `pureGetters` assumption looks good to me. Do you want to open a PR?

**Comment by Austaras** [comment]
Yes

**Comment by Austaras** [comment]
What do I need to do if I want to add pureGetters options to nullish coalescing? I don't seem to find a guide

**Comment by Austaras** [comment]
> Respecting the `pureGetters` assumption looks good to me. Do you want to open a PR?
> 
> The `pureGetters` assumption can not cover the case, even if the getter is pure. The fact that everytime when `foo` is accessed, a new object is returned does prove that it should be memoized.

That would apply to the docuemented case too consider

```
let a = { get b() { return function() {} } };

a.b?.();
```

since every access to b will cause a new function to be constructed

**Comment by nicolo-ribaudo** [comment]
Yeah, I agree that this can go behind `pureGetters`. The common definition of "pure function" is a function that:
- Given the same arguments return the same result
- Doesn't have side effects

`let a = { get b() { return function() {} } };` can be considered as non-pure, because `a.b !== a.b` (it violates the first point, if we define "same" as `===`).

**Comment by JLHwung** [comment]
Maybe we can introduce a new assumption like `noGlobalThisGetters`?

**Comment by Austaras** [comment]
Another question: should long optional chain like `a?.b?.c?.d` be cached?

**Comment by The-x-Theorist** [comment]
> Maybe we can introduce a new assumption like `noGlobalThisGetters`?

Normally maximum Getters will be not globalThis, Making an assumption only for globalGetters can make more sense?
